### ISSUE TITLE
Wired behavior for spack concretizer

### ISSUE THREAD

**Issue by Guangzhen-Jin** [LLNL/spack#45510]
### Steps to reproduce

I am using spack environment as follows:

``` console
spack:
  # add package specs to the `specs` list
  definitions:
  - compilers:
    - gcc@8.5.0
    - gcc@11.2.0
    - gcc@14.1.0
  - generic_packages:
    - netlib-lapack@3.11.0
    - fftw@3.3.10 ~mpi
    - fftw@2.1.5 ~mpi
    - openblas@0.3.27
    - boost@1.85.0 ~mpi
    - hdf@4.2.15
    - hdf5@1.14.3 ~mpi
    - netcdf-c@4.9.2 ~mpi
    - netcdf-fortran@4.6.1 ^netcdf-c@4.9.2~mpi
    - netcdf-cxx4@4.3.1 ^netcdf-c@4.9.2~mpi ^hdf5@1.14.3 ~mpi
  specs:
  - matrix:
    - [$generic_packages]
    - [$%compilers]
    view: false
  concretizer:
    unify: when_possible

  modules:
    default:
      lmod:
        core_compilers:
        - gcc@=8.5.0
```

The installation went without issues but `spack spec` indicates the dependents for the packages were built with different compilers (see below an example), which will cause issues while trying to generate modules for use.

``` console
$ spack spec boost@1.85.0%gcc@8.5.0
Input spec
--------------------------------
 -   boost@1.85.0%gcc@8.5.0

Concretized
--------------------------------
[+]  boost@1.85.0%gcc@8.5.0~atomic~chrono~clanglibcpp~container~context~contract~coroutine~date_time~debug~exception~fiber~filesystem~graph~graph_parallel~icu~iostreams~json~locale~log~math~mpi+multithreaded~nowide~numpy~pic~program_options~python~random~regex~serialization+shared~signals~singlethreaded~stacktrace~system~taggedlayout~test~thread~timer~type_erasure~versionedlayout~wave build_system=generic cxxstd=11 patches=a440f96 visibility=hidden arch=linux-rocky8-zen
[+]      ^gcc-runtime@14.1.0%gcc@14.1.0 build_system=generic arch=linux-rocky8-zen
[e]          ^glibc@2.28%gcc@11.2.0 build_system=autotools arch=linux-rocky8-zen
[e]      ^glibc@2.28%gcc@8.5.0 build_system=autotools arch=linux-rocky8-zen
```

Anyone can help on this? Thank you.

### Error message

No error message until when you try to load the module. I get the reason should be the same as mentioned above.

``` console
$ module load boost
Lmod has detected the following error:  These module(s) or extension(s) exist but cannot be loaded as requested: "gcc-runtime/14.1.0-tt2d5iw"
   Try: "module spider gcc-runtime/14.1.0-tt2d5iw" to see how to load the module(s).

Executing this command requires loading "gcc-runtime/14.1.0-tt2d5iw" which failed while processing the following module(s):

    Module fullname  Module Filename
    ---------------  ---------------
    boost/1.85.0     /opt/spack/cpu/Core/boost/1.85.0.lua
```

### Information on your system

* **Spack:** 0.23.0.dev0 (f99a5ef2e7afd001d3ee6ca1bbb662c2f61903b8)
* **Python:** 3.6.8
* **Platform:** linux-rocky8-zen2
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
A few comments:
1. Using `unify: when_possible` encourages mixing compilers, and doesn't make much sense in the case above with a matrix like yours
2. The compiler in the matrix is effectively enforced only on the root node
3. You are likely reusing specs you have installed, or in some registered buildcache. I guess that from the fact that you have dependents using a gcc with a lower version wrt dependencies. That should be fixed on `develop`.

**TL;DR**, if you use:
```console
concretizer:
  unify: false
  reuse: false
```
that should have the effect you desire (compile each spec separately, with a single compiler).
